applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nonlinear_strain_energy_response_function.cpp

```diff
@@ -0,0 +1,378 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Mahmoud Sesa, https://github.com/mahmoudsesa
+//
+
+// System includes
+
+// External includes
+#include "utilities/variable_utils.h"
+
+// Project includes
+#include "adjoint_nonlinear_strain_energy_response_function.h"
+
+namespace Kratos 
+{
+    AdjointNonlinearStrainEnergyResponseFunction::AdjointNonlinearStrainEnergyResponseFunction(ModelPart& rModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rModelPart, ResponseSettings)
+    {
+        // initializing mConditionsRHS 
+        for(auto& cond_i : rModelPart.Conditions())
+        {
+            SizeType number_of_nodes = cond_i.GetGeometry().size();
+            SizeType dimension = cond_i.GetGeometry().WorkingSpaceDimension();
+            SizeType vec_size = number_of_nodes * dimension;
+          
+            mConditionsRHS[cond_i.Id()] = ZeroVector(vec_size);
+
+            // This stores a vector of pointers to the model condition as a member variable because I can not use const condition in 
+            mConditions[cond_i.Id()] = rModelPart.pGetCondition(cond_i.Id());
```

Comment:
this is a find, i.e. expensive esp for large models

Reply:
Yes, the problem is that in function CalculateGradient() the condition is defined as const, this does not allow me to calculate the condition derivative. const should be removed from the base class as in CalculatePartialSensitivity().

Reply:
I dont understand :)

I mean you should not use `pGetCondition` if possible
However this is more a comment towards @armingeiser and @MFusseder

Reply:
I agree the problem is that they did not consider my case when they have written the base class, so till they change it I could not find out another way.

Reply:
As an intermediate step you could loop using the iterators `ConditionsBegin(), ConditionsEnd()`. Then you would use the pointers directly and don't need to call `pGetCondition` additionally

Reply:
okay, I will try this.